File: src/mpi_laplace.cpp
#include <mpi.h>
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <iomanip>

using namespace std;

int main(int argc, char* argv[]) {
    MPI_Init(&argc, &argv);

    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    int N = 200;
    int max_iter = 10000;
    double tol = 1e-6;

    if (argc > 1) N = stoi(argv[1]);
    if (argc > 2) max_iter = stoi(argv[2]);
    if (argc > 3) tol = stod(argv[3]);

    int base_rows = N / size;
    int extra = N % size;
    int local_rows = base_rows + (rank < extra ? 1 : 0);

    int start_row = rank * base_rows + min(rank, extra);

    vector<vector<double>> u(local_rows + 2, vector<double>(N, 0.0));
    vector<vector<double>> u_new(local_rows + 2, vector<double>(N, 0.0));

    // Top boundary condition = 100
    for (int i = 1; i <= local_rows; ++i) {
        int global_i = start_row + (i - 1);
        if (global_i == 0) {
            for (int j = 0; j < N; ++j) {
                u[i][j] = 100.0;
                u_new[i][j] = 100.0;
            }
        }
    }

    MPI_Barrier(MPI_COMM_WORLD);
    double total_start = MPI_Wtime();

    int iter = 0;
    double global_diff = 0.0;
    double comm_time = 0.0;
    double update_time = 0.0;
    double copy_time = 0.0;
    bool converged = false;

    for (iter = 0; iter < max_iter; ++iter) {
        double comm_start = MPI_Wtime();

        if (rank > 0) {
            MPI_Sendrecv(
                u[1].data(), N, MPI_DOUBLE, rank - 1, 0,
                u[0].data(), N, MPI_DOUBLE, rank - 1, 1,
                MPI_COMM_WORLD, MPI_STATUS_IGNORE
            );
        }

        if (rank < size - 1) {
            MPI_Sendrecv(
                u[local_rows].data(), N, MPI_DOUBLE, rank + 1, 1,
                u[local_rows + 1].data(), N, MPI_DOUBLE, rank + 1, 0,
                MPI_COMM_WORLD, MPI_STATUS_IGNORE
            );
        }

        double comm_end = MPI_Wtime();
        comm_time += (comm_end - comm_start);

        double local_diff = 0.0;

        double update_start = MPI_Wtime();
        for (int i = 1; i <= local_rows; ++i) {
            int global_i = start_row + (i - 1);
            if (global_i == 0 || global_i == N - 1) continue;

            for (int j = 1; j < N - 1; ++j) {
                u_new[i][j] = 0.25 * (
                    u[i - 1][j] +
                    u[i + 1][j] +
                    u[i][j - 1] +
                    u[i][j + 1]
                );

                double diff = fabs(u_new[i][j] - u[i][j]);
                if (diff > local_diff) {
                    local_diff = diff;
                }
            }
        }
        double update_end = MPI_Wtime();
        update_time += (update_end - update_start);

        double copy_start = MPI_Wtime();
        for (int i = 1; i <= local_rows; ++i) {
            int global_i = start_row + (i - 1);
            if (global_i == 0 || global_i == N - 1) continue;

            for (int j = 1; j < N - 1; ++j) {
                u[i][j] = u_new[i][j];
            }
        }
        double copy_end = MPI_Wtime();
        copy_time += (copy_end - copy_start);

        MPI_Allreduce(&local_diff, &global_diff, 1, MPI_DOUBLE, MPI_MAX, MPI_COMM_WORLD);

        if (global_diff < tol) {
            converged = true;
            break;
        }
    }

    double total_end = MPI_Wtime();
    double total_time = total_end - total_start;
    int iterations_reported = converged ? iter + 1 : iter;

    vector<int> recvcounts(size), displs(size);
    for (int r = 0; r < size; ++r) {
        recvcounts[r] = (base_rows + (r < extra ? 1 : 0)) * N;
        displs[r] = (r == 0) ? 0 : displs[r - 1] + recvcounts[r - 1];
    }

    vector<double> sendbuf(local_rows * N);
    for (int i = 0; i < local_rows; ++i) {
        for (int j = 0; j < N; ++j) {
            sendbuf[i * N + j] = u[i + 1][j];
        }
    }

    vector<double> fullgrid;
    if (rank == 0) {
        fullgrid.resize(N * N);
    }

    MPI_Gatherv(
        sendbuf.data(), local_rows * N, MPI_DOUBLE,
        fullgrid.data(), recvcounts.data(), displs.data(), MPI_DOUBLE,
        0, MPI_COMM_WORLD
    );

    if (rank == 0) {
        cout << fixed << setprecision(6);
        cout << "Grid size: " << N << "x" << N << "\n";
        cout << "MPI processes: " << size << "\n";
        cout << "Iterations: " << iterations_reported << "\n";
        cout << "Final max diff: " << global_diff << "\n";
        cout << "Center value: " << fullgrid[(N / 2) * N + (N / 2)] << "\n";
        cout << "Converged: " << (converged ? "true" : "false") << "\n";
        cout << "Total time: " << total_time << " seconds\n";
        cout << "Communication time: " << comm_time << " seconds\n";
        cout << "Update loop time: " << update_time << " seconds\n";
        cout << "Copy loop time: " << copy_time << " seconds\n";

        ofstream fout("results/solution_csv/cpp_mpi_solution.csv");
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                fout << fullgrid[i * N + j];
                if (j != N - 1) fout << ",";
            }
            fout << "\n";
        }
        fout.close();
    }

    MPI_Finalize();
    return 0;
}

File: src/mpi_laplace.py
from mpi4py import MPI
import numpy as np
import sys

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

N = int(sys.argv[1]) if len(sys.argv) > 1 else 200
max_iter = int(sys.argv[2]) if len(sys.argv) > 2 else 10000
tol = float(sys.argv[3]) if len(sys.argv) > 3 else 1e-6

base_rows = N // size
extra = N % size
local_rows = base_rows + (1 if rank < extra else 0)
start_row = rank * base_rows + min(rank, extra)

u = np.zeros((local_rows + 2, N), dtype=np.float64)
u_new = np.zeros_like(u)

# Top boundary = 100
for i in range(1, local_rows + 1):
    global_i = start_row + (i - 1)
    if global_i == 0:
        u[i, :] = 100.0
        u_new[i, :] = 100.0

comm.Barrier()
total_start = MPI.Wtime()

comm_time = 0.0
update_time = 0.0
copy_time = 0.0
converged = False
final_diff = None

for it in range(max_iter):
    t0 = MPI.Wtime()

    if rank > 0:
        comm.Sendrecv(sendbuf=u[1, :], dest=rank - 1, sendtag=0,
                      recvbuf=u[0, :], source=rank - 1, recvtag=1)

    if rank < size - 1:
        comm.Sendrecv(sendbuf=u[local_rows, :], dest=rank + 1, sendtag=1,
                      recvbuf=u[local_rows + 1, :], source=rank + 1, recvtag=0)

    t1 = MPI.Wtime()
    comm_time += (t1 - t0)

    local_diff = 0.0

    t2 = MPI.Wtime()
    for i in range(1, local_rows + 1):
        global_i = start_row + (i - 1)
        if global_i == 0 or global_i == N - 1:
            continue

        u_new[i, 1:-1] = 0.25 * (
            u[i - 1, 1:-1] +
            u[i + 1, 1:-1] +
            u[i, :-2] +
            u[i, 2:]
        )

        diff = np.max(np.abs(u_new[i, 1:-1] - u[i, 1:-1]))
        local_diff = max(local_diff, diff)
    t3 = MPI.Wtime()
    update_time += (t3 - t2)

    t4 = MPI.Wtime()
    for i in range(1, local_rows + 1):
        global_i = start_row + (i - 1)
        if global_i == 0 or global_i == N - 1:
            continue
        u[i, 1:-1] = u_new[i, 1:-1]
    t5 = MPI.Wtime()
    copy_time += (t5 - t4)

    global_diff = comm.allreduce(local_diff, op=MPI.MAX)
    final_diff = global_diff

    if global_diff < tol:
        converged = True
        break

total_end = MPI.Wtime()
total_time = total_end - total_start
iterations_reported = (it + 1) if converged else it

sendbuf = u[1:local_rows + 1, :].ravel()
recvcounts = np.array(
    [(base_rows + (1 if r < extra else 0)) * N for r in range(size)],
    dtype=int
)
displs = np.zeros(size, dtype=int)
displs[1:] = np.cumsum(recvcounts[:-1])

if rank == 0:
    fullgrid = np.empty(N * N, dtype=np.float64)
else:
    fullgrid = None

comm.Gatherv(sendbuf, [fullgrid, recvcounts, displs, MPI.DOUBLE], root=0)

if rank == 0:
    fullgrid = fullgrid.reshape((N, N))
    print(f"Grid size: {N}x{N}")
    print(f"MPI processes: {size}")
    print(f"Iterations: {iterations_reported}")
    print(f"Final max diff: {final_diff:.6e}")
    print(f"Center value: {fullgrid[N // 2, N // 2]:.6f}")
    print(f"Converged: {converged}")
    print(f"Total time: {total_time:.6f} seconds")
    print(f"Communication time: {comm_time:.6f} seconds")
    print(f"Update loop time: {update_time:.6f} seconds")
    print(f"Copy loop time: {copy_time:.6f} seconds")

    np.savetxt("results/solution_csv/python_mpi_solution.csv", fullgrid, delimiter=",", fmt="%.6f")
